packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -473,6 +473,18 @@ abstract class Widget extends DiagnosticableTree {
     return oldWidget.runtimeType == newWidget.runtimeType
         && oldWidget.key == newWidget.key;
   }
+
+  // Return a numeric encoding of the specific `Widget` concrete subtype.
+  // This is used to determine if a hot reload modified the type hierarchy of
+  // a mounted element. The encoding of each `Widget` must match the corresponding
+  // `Element` encoding in `Element._debugConcreteSubtype`.
+  static int _debugConcreteSubtype(Widget widget) {
+    return widget is StatelessWidget ? 1 :
+           widget is StatefulWidget ? 2 :
+           widget is SingleChildRenderObjectWidget ? 3 :
+           widget is InheritedWidget? 4 :
+           0;
```

Comment:
we should just add all the common subclasses, sorry that wasn't clear in my earlier comment. These:
StatefulElement, StatelessElement, InheritedModelElement, InheritedElement, ParentDataElement, ProxyElement, LeafRenderObjectElement, RootRenderObjectElement, SingleChildRenderObjectElement, MultiChildRenderObjectElement, RenderObjectElement. (In that order, since some are superclasses of each other.)

Reply:
I've updated with the types here, save for RootRenderObjectElement which uses a regular RenderObjectWidget as its configuration